src/wp-includes/option.php

```diff
@@ -253,6 +253,110 @@ function get_option( $option, $default_value = false ) {
 	return apply_filters( "option_{$option}", maybe_unserialize( $value ), $option );
 }
 
+/**
+ * Primes specific options into the cache with a single database query.
+ *
+ * Only options that do not already exist in cache will be primed.
+ *
+ * @since 6.4.0
+ *
+ * @global wpdb $wpdb WordPress database abstraction object.
+ *
+ * @param array $options An array of option names to be primed.
+ */
+function prime_options( $options ) {
+	$alloptions     = wp_load_alloptions();
+	$cached_options = wp_cache_get_multiple( $options, 'options' );
+
+	// Filter options that are not in the cache.
+	$options_to_prime = array();
+	foreach ( $options as $option ) {
+		if ( ( ! isset( $cached_options[ $option ] ) || ! $cached_options[ $option ] ) && ! isset( $alloptions[ $option ] ) ) {
+			$options_to_prime[] = $option;
+		}
+	}
+
+	// Bail early if there are no options to be primed.
+	if ( empty( $options_to_prime ) ) {
+		return;
+	}
+
+	global $wpdb;
+	$results = $wpdb->get_results(
+		$wpdb->prepare(
+			sprintf(
+				"SELECT option_name, option_value FROM $wpdb->options WHERE option_name IN (%s)",
+				implode( ',', array_fill( 0, count( $options_to_prime ), '%s' ) )
+			),
+			$options_to_prime
+		)
+	);
+
+	$options_not_found = $options_to_prime;
+	foreach ( $results as $result ) {
+		wp_cache_set( $result->option_name, maybe_unserialize( $result->option_value ), 'options' );
+		unset( $options_not_found[ array_search( $result->option_name, $options_not_found, true ) ] );
+	}
```

Comment:
This logic can be optimized, to avoid the use of `array_search` which is more expensive than e.g. `isset()`, and to use `wp_cache_set_multiple()` instead of individual `wp_cache_set()` calls. Additionally, if all options were found, we don't need to run any of the logic further below.
```suggestion
	$options_found = array();
	foreach ( $results as $result ) {
		$options_found[ $result->option_name ] =  maybe_unserialize( $result->option_value );
	}
	wp_cache_set_multiple( $options_found, 'options' );

	// If all options were found, no need to update `notoptions` cache.
	if ( count( $options_found ) === count( $options_to_prime ) ) {
		return;
	}

	$options_not_found = array_diff( $options_to_prime, array_keys( $options_found ) );
```

Reply:
@felixarntz `wp_cache_set_multiple` ultimately loops through these a second time [here](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-object-cache.php#L337-L339), so by calling `wp_cache_set()` directly here, it should mean looping once in total, rather than once here, and once in `WP_Object_Cache::set_multiple()`. What are your thoughts on this?

Regarding the removal of `array_search()`, the change to `array_diff( array_keys() )` looks good to me, avoiding `array_search()` per `foreach()` iteration.

Reply:
A second loop doesn't really make a difference in performance, however more importantly what you're describing is only the _default_ core implementation of `wp_cache_set_multiple()`. There are persistent object cache engines that truly allow multi-setting values, that's why that function was introduced in the first place.

In other words: Wherever we can use `wp_cache_set_multiple()` on groups of values in the same cache group, we should, as it allows to use those benefits where available, while not having a real drawback otherwise (again, the second `foreach` doesn't make any real performance difference).

Reply:
Thanks for the clarification Felix! 🙂